lib/widgets/compose_box.dart

```diff
@@ -257,6 +317,91 @@ class _StreamSendButtonState extends State<_StreamSendButton> {
   }
 }
 
+class _AttachFileButton extends StatelessWidget {
+  const _AttachFileButton({required this.contentController});
+
+  final ContentTextEditingController contentController;
+
+  _handlePress(BuildContext context) async {
+    FilePickerResult? result;
+    try {
+      result = await FilePicker.platform.pickFiles(allowMultiple: true, withReadStream: true);
+    } catch (e) {
+      // TODO(i18n)
+      showErrorDialog(context: context, title: 'Error', message: e.toString());
+      return;
+    }
+
+    if (result == null) {
+      return; // User cancelled; do nothing
+    }
+
+    if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
+    else {
+      return;
+    }
+
+    final store = PerAccountStoreWidget.of(context);
+    final account = store.account;
+    final realmUrl = Uri.parse(account.realmUrl);
+
+    final List<PlatformFile> tooLargeFiles = [];
+    final List<PlatformFile> rightSizeFiles = [];
+
+    for (PlatformFile file in result.files) {
+      if ((file.size / pow(2, 20)) > store.maxFileUploadSizeMib) {
+        tooLargeFiles.add(file);
+      } else {
+        rightSizeFiles.add(file);
+      }
+    }
+
+
+    if (tooLargeFiles.isNotEmpty) {
+      final listMessage = tooLargeFiles
+        .map((file) => '${file.name}: ${(file.size / pow(2, 20)).toStringAsFixed(1)} MiB')
+        .join('\n');
+      showErrorDialog( // TODO(i18n)
+        context: context,
+        title: 'File(s) too large',
+        message:
+          '${tooLargeFiles.length} file(s) are larger than the server\'s limit of ${store.maxFileUploadSizeMib} MiB and will not be uploaded:\n\n$listMessage');
+    }
+
+    final List<(int, PlatformFile)> uploadsInProgress = [];
+    for (final file in rightSizeFiles) {
+      final tag = contentController.registerUploadStart(file.name);
+      uploadsInProgress.add((tag, file));
+    }
+
+    for (final (tag, file) in uploadsInProgress) {
+      final readStream = file.readStream;
+      final size = file.size;
+      final name = file.name;
+
+      // Will fail if we didn't pass `withReadStream: true` to pickFiles
+      assert(readStream != null, 'readStream missing');
+      Uri? url;
+      try {
+        final result = await uploadFile(store.connection, content: readStream!, length: size, filename: name);
+        url = realmUrl.resolve(result.uri);
+      } catch (e) {
+        if (!context.mounted) return;
+        // TODO: Specifically handle `413 Payload Too Large`
+        // TODO: On API errors, quote `msg` from server, with "The server said:"
```

Comment:
Can help make these findable when we do the underlying work to make them possible:

```suggestion
        // TODO(#37): Specifically handle `413 Payload Too Large`
        // TODO(#37): On API errors, quote `msg` from server, with "The server said:"
```

Reply:
Oh right; I'd forgotten we had that issue #37, thanks.